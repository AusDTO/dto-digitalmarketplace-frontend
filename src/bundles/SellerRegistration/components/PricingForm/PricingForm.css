.inline-input input{
  width: 12%;
  margin-left: 5px;
  margin-right: 5px;
}
