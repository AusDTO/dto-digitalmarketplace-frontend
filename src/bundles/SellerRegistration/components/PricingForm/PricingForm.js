import React                from 'react';
import PropTypes            from 'prop-types'
import { connect }          from 'react-redux';
import { Link }             from 'react-router-dom';
import { Form, Control }    from 'react-redux-form';
import isEmpty              from 'lodash/isEmpty';
import kebabCase            from 'lodash/kebabCase';

import Layout               from '../../../../shared/Layout';
import BaseForm             from '../../../../shared/form/BaseForm';
import ErrorBox             from '../../../../shared/form/ErrorBox';
import StatefulError        from '../../../../shared/form/StatefulError';
import formProps            from '../../../../shared/reduxModules/formPropsSelector';
import SubmitForm           from '../../../../shared/form/SubmitForm';

import StepNav              from '../StepNav';
import { findValidServices } from '../../redux/helpers';
import { required, notNegativeNumber, onlyWholeNumbers }         from '../../../../validators';

import PageAlert from '@gov.au/page-alerts'

import styles from '../SellerRegistration.css';
import pricing from './PricingForm.css';

class PricingForm extends BaseForm {

  static propTypes = {
    model: PropTypes.string.isRequired,
    form: PropTypes.object.isRequired,
    action: PropTypes.string.isRequired,
    csrf_token: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    buttonText: PropTypes.string.isRequired,
    services: PropTypes.object.isRequired,
    children: PropTypes.object.isRequired,
    onSubmit: PropTypes.func.isRequired,
    nextRoute: PropTypes.string.isRequired,
    submitClicked: PropTypes.bool,
  }

  render() {
    const { model, form, action, csrf_token, title, buttonText, services, children,  onSubmit, nextRoute, submitClicked, domains, applicationErrors } = this.props;
    let validServices = findValidServices(services);

    if (isEmpty(validServices)) {
      return (
        <Layout>
          <header>
            <h1 className="au-display-xl" tabIndex="-1">{title}</h1>
          </header>
          <article role="main">
            <p>You must select the services you offer before specifying pricing!</p>
            <p><Link to="/domains">Select services</Link></p>
          </article>
        </Layout>
      )
    }

    let hasFocused = false
    const setFocus = e => {
      if (!hasFocused) {
        hasFocused = true
        e.focus()
      }
    }

    return (
      <Layout>
        <header styleName="styles.content">
            { (form.submitFailed === false) && applicationErrors.length > 0 ? (
              <PageAlert as="error">
                  <h3>Errors</h3>
                  <ul>
                  {applicationErrors.map(ae => {
                      return <li key="{ae.message}">{ae.message}</li>
                  })}
                  </ul>
              </PageAlert> ) : ''
            }
            <h1 className="au-display-xl" styleName="styles.content-heading" tabIndex="-1">{title}</h1>
            <p>
              The Marketplace asks for a comparable level of pricing across all sellers. This helps ensure that your 
              business offers value for money.
            </p>
            <p>
              To do this, we ask sellers to list their maximum daily rate (including GST) for each service, using {' '}
              <a href="https://www.sfia-online.org/en/sfia-6/busskills/lr5" rel="external nofollow">
                 Skills Framework for the Information Age (SFIA) level 5
              </a>as a guide.
            </p>
            <p>
              For more information, you can also read our {' '}
              <a href="https://marketplace1.zendesk.com/hc/en-gb/articles/360000054595-Seller-pricing" rel="external">
                 Seller Pricing FAQs
              </a>
              and view the 
              <a href="https://marketplace1.zendesk.com/hc/en-gb/articles/360000080555-Daily-rates-trend-charts" rel="external">
                daily rates trend charts
              </a>.
            </p>
        </header>
        <article role="main">
          <ErrorBox submitClicked={submitClicked} model={model} setFocus={setFocus}/>
          <Form
            model={model}
            action={action}
            method="post"
            id="DomainSelector__create"
            component={SubmitForm}
            onCustomSubmit={onSubmit}
            valid={form.valid}
            >
            {csrf_token && (
              <input type="hidden" name="csrf_token" id="csrf_token" value={csrf_token}/>
            )}

            {Object.keys(validServices).map((service, i) => (
              <fieldset key={`pricing.${service}.${i}`} className="field">
                <legend>{service}</legend>
                <label>
                  We have set a price threshold for this area of expertise: {`$${parseFloat(domains.prices.maximum[service] || 0)}`} (GST inclusive). If you nominate a price above this threshold we will assess your case studies at a higher standard to verify that your service offers value for money. 
                </label>
                <StatefulError
                  model={`${model}.pricing.${service}.maxPrice`}
                  messages={{
                    required: `You must provide a max price for ${service}.`,
                    notNegativeNumber: `Price can't have negative number for ${service}.`,
                    onlyWholeNumbers: `Price must be whole dollars for ${service}.`
                  }}
                  id={`${kebabCase(service)}-maxprice`}
                />
                <div>
                  <span><strong>$</strong></span>
                  <span styleName="pricing.inline-input">
                    <Control.text
                      type="text"
                      id={`${kebabCase(service)}-maxprice`}
                      name={`services.${service}.maxPrice`}
                      model={`${model}.pricing.${service}.maxPrice`}
                      validators={{ required, notNegativeNumber, onlyWholeNumbers }}
                    />
                  </span>
                  <span>(including GST)</span>
                </div>
              </fieldset>
            ))}

            {children}

            <StepNav buttonText={buttonText} to={nextRoute}/>
          </Form>
        </article>
      </Layout>
    )
  }
}

PricingForm.defaultProps = {
  buttonText: 'Save &amp; Continue',
  title: 'Pricing',
  services: {}
}

const mapStateToProps = (state) => {
  return {
    ...formProps(state, 'pricingForm'),
    domainSelectorForm: state.domainSelectorForm,
    applicationErrors: state.application_errors.filter(ae => ae.step === 'pricing')
  }
}

export default connect(mapStateToProps)(PricingForm)
