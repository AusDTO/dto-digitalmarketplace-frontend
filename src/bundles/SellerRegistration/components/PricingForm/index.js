import PricingForm from './PricingForm';

export default PricingForm;