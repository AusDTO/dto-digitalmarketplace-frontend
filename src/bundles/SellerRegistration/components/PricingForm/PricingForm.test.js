// FIXME This is a workaround.
// @see https://github.com/facebook/react/issues/7386
jest.mock('react-dom');

import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureMockStore from 'redux-mock-store';

import { Router } from 'react-router-dom';
import createMemoryHistory from 'history/createMemoryHistory'

import PricingForm from './PricingForm'
import createStore from '../../redux/create-signup'

test('PricingForm renders a notice that domains are required', () => {
  const store = createStore({
    form_options: {
      csrf_token: 'abc'
    },
    options: {
      serverRender: false
    }
  });

  const history = createMemoryHistory();
  const component = renderer.create(
    <Router history={history}>
      <Provider store={store}>
        <PricingForm services={{}} />
      </Provider>
    </Router>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('PricingForm renders price fields when domains are present', () => {
  const domains = {
    'User research': 123,
    'Content development': 456,
    'Content management': 789
  }

  const services = {
    'User research': true,
    'Content development': true,
    'Content management': true
  }
  const store = createStore({
    form_options: {
      csrf_token: 'abc'
    },
    options: {
      serverRender: false
    }
  });

  const component = renderer.create(
    <Provider store={store}>
      <PricingForm domains={domains} services={services} />
    </Provider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

