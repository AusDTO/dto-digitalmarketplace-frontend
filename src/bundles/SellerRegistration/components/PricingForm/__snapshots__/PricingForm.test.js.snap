// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PricingForm renders a notice that domains are required 1`] = `
<div
  className="layout"
>
  <header>
    <h1
      tabIndex="-1"
    >
      Pricing
    </h1>
  </header>
  <article
    role="main"
  >
    <p>
      You must select some domains before specifying pricing!
    </p>
    <p>
      <a
        href="/domains"
        onClick={[Function]}
      >
        Select domains
      </a>
    </p>
  </article>
</div>
`;

exports[`PricingForm renders price fields when domains are present 1`] = `
<div
  className="layout"
>
  <header>
    <h1
      tabIndex="-1"
    >
      Pricing
    </h1>
    <p>
      This will only be visible to government buyers, not to other sellers on the Marketplace.
    </p>
  </header>
  <article
    role="main"
  >
    <form
      action={undefined}
      id="DomainSelector__create"
      method="post"
      noValidate={undefined}
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <input
        id="csrf_token"
        name="csrf_token"
        type="hidden"
        value="abc"
      />
      <fieldset
        className="field"
      >
        <legend>
          User research
        </legend>
        <div
          className="row"
        >
          <div
            className="col-sm-5 col-xs-12"
          >
            <label
              htmlFor="user-research-minprice"
            >
              Min. Daily Price
            </label>
            <input
              disabled={false}
              id="user-research-minprice"
              name="services.User research.minPrice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              type="number"
              value=""
            />
          </div>
          <div
            className="col-sm-push-1 col-sm-5 col-xs-12"
          >
            <label
              htmlFor="user-research-maxprice"
            >
              Max. Daily Price
            </label>
            <input
              disabled={false}
              id="user-research-maxprice"
              name="services.User research.maxPrice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              type="number"
              value=""
            />
          </div>
        </div>
      </fieldset>
      <fieldset
        className="field"
      >
        <legend>
          Content development
        </legend>
        <div
          className="row"
        >
          <div
            className="col-sm-5 col-xs-12"
          >
            <label
              htmlFor="content-development-minprice"
            >
              Min. Daily Price
            </label>
            <input
              disabled={false}
              id="content-development-minprice"
              name="services.Content development.minPrice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              type="number"
              value=""
            />
          </div>
          <div
            className="col-sm-push-1 col-sm-5 col-xs-12"
          >
            <label
              htmlFor="content-development-maxprice"
            >
              Max. Daily Price
            </label>
            <input
              disabled={false}
              id="content-development-maxprice"
              name="services.Content development.maxPrice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              type="number"
              value=""
            />
          </div>
        </div>
      </fieldset>
      <fieldset
        className="field"
      >
        <legend>
          Content management
        </legend>
        <div
          className="row"
        >
          <div
            className="col-sm-5 col-xs-12"
          >
            <label
              htmlFor="content-management-minprice"
            >
              Min. Daily Price
            </label>
            <input
              disabled={false}
              id="content-management-minprice"
              name="services.Content management.minPrice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              type="number"
              value=""
            />
          </div>
          <div
            className="col-sm-push-1 col-sm-5 col-xs-12"
          >
            <label
              htmlFor="content-management-maxprice"
            >
              Max. Daily Price
            </label>
            <input
              disabled={false}
              id="content-management-maxprice"
              name="services.Content management.maxPrice"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              type="number"
              value=""
            />
          </div>
        </div>
      </fieldset>
      <input
        type="submit"
        value="Save &amp; Continue"
      />
    </form>
  </article>
</div>
`;
